isic/core/api/collection.py

```diff
@@ -1,105 +1,125 @@
-        )
-        return Response(status=status.HTTP_202_ACCEPTED)
-
-    # TODO: refactor *-from-list methods
-    @swagger_auto_schema(auto_schema=None)
-    @action(detail=True, methods=["post"], pagination_class=None, url_path="populate-from-list")
-    def populate_from_list(self, request, *args, **kwargs):
-        if not request.user.has_perm("core.add_images", self.get_object()):
-            raise PermissionDenied
-
-        self._enforce_write_checks(request.user)
-        serializer = IsicIdListSerializer(data=request.data)
-        serializer.is_valid(raise_exception=True)
-
-        summary = collection_add_images_from_isic_ids(
-            user=request.user,
-            collection=self.get_object(),
-            isic_ids=serializer.validated_data["isic_ids"],
-        )
-
-        return JsonResponse(summary)
-
-    @swagger_auto_schema(auto_schema=None)
-    @action(detail=True, methods=["post"], pagination_class=None, url_path="remove-from-list")
-    def remove_from_list(self, request, *args, **kwargs):
-        if not request.user.has_perm("core.remove_images", self.get_object()):
-            raise PermissionDenied
-
-        self._enforce_write_checks(request.user)
-        serializer = IsicIdListSerializer(data=request.data)
-        serializer.is_valid(raise_exception=True)
-
-        summary = collection_remove_images_from_isic_ids(
-            user=request.user,
-            collection=self.get_object(),
-            isic_ids=serializer.validated_data["isic_ids"],
-        )
-
-        # TODO: this is a weird mixture of concerns between SSR and an API, figure out a better
-        # way to handle this.
-        messages.add_message(request, messages.INFO, f'Removed {len(summary["succeeded"])} images.')
-
-        return JsonResponse(summary)
+class CollectionOut(ModelSchema):
+    class Config:
+        model = Collection
+        model_fields = ["id", "name", "description", "public", "pinned", "locked", "doi"]
+
+    doi_url: str | None
+
+    @staticmethod
+    def resolve_doi_url(obj: Collection):
+        return obj.doi_url
+
+
+@router.get("/", response=list[CollectionOut])
+@paginate(CursorPagination)
+def collection_list(request, pinned: bool | None = None) -> list[CollectionOut]:
+    queryset = get_visible_objects(request.user, "core.view_collection", Collection.objects.all())
+    if pinned is not None:
+        queryset = queryset.filter(pinned=pinned)
+    return queryset
+
+
+@router.get("/{id}/", response=CollectionOut)
+def collection_detail(request, id: int) -> CollectionOut:
+    qs = get_visible_objects(request.user, "core.view_collection", Collection.objects.all())
+    collection = get_object_or_404(qs, id=id)
+    return collection
+
+
+@router.post("/{id}/populate-from-search/", response={202: None, 403: dict, 409: dict})
+def collection_populate_from_search(request, id: int, payload: SearchQueryIn):
+    collection = get_object_or_404(Collection, id=id)
```

Comment:
Is this behaving identically? I thought `self.get_object()` did permissions checks but I can't remember.

Reply:
Ah, good catch. And this is another good reason to move away from DRF CBVs :)

Reply:
Do you think it's worth adding something in the `CollectionQuerySet` manager like `def get_visible_to(user)`, or does that complicate the model layer too much?

Reply:
I don't think it's worth doing, at least not right now. We would need them on almost every model. The future plan is to use guardian everywhere (https://github.com/ImageMarkup/tracker/issues/77).